ratex/core/distributed/ratex_fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,85 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Module wrapper implementing ZeRO-1 in an FSDP style interface"""
+import torch.nn as nn
+import torch.nn.functional as F
+
+from raf import distributed as dist
+
+import ratex.optimizer as optim
+from ratex.core.lazy_model import all_gather, all_reduce, REDUCE_SUM
+
+
+class RatexFullyShardedDataParallel(nn.Module):
+    """FSDP ZeRO-1 wrapper"""
+
+    def __init__(self, module, optimizer=optim.SGD, optimizer_config=None):
+        super().__init__()
+        comm = dist.get_communicator()
+        self.rank = comm.rank
+        self.world_size = comm.size
+        self.module = module
+        self.params = list(self.module.parameters())
+
+        # Shard parameters for use in optimizer
+        self._shard_parameters()
+        # Optimizer initialization
+        self.optimizer = optimizer(self.sharded_parameters(), **optimizer_config or {})
+
+    def _shard_parameters(self):
+        """
+        Create list of this ranks shards
+        Shards are stored in a tuple with the respective parameter ie (parameter, shard)
+        """
+        self.sharded_params = []
```

Comment:
1. Avoid creating class members other than constructor. You could put this line in `__init__` instead.
2. Why we need to shard parameters in ZeRO-1?

Reply:
I changed the initialization of sharded_params to the constructor. The parameters are sharded in this implementation to be passed to the optimizer. This way each rank only stares shards of the optimizer states.

Reply:
I got your point. You use shard parameters to create optimizer, so that optimizer states are also sharded. The shard parameters are parts of the optimizer states.